WasatchNET/Spectrometer.cs

```diff
@@ -1814,29 +1863,83 @@ virtual internal bool open()
+            logger.debug($"Spectrometer.close: closing {id}");
+
+            // quit whatever we're doing
+            cancelCurrentAcquisition();
+
+            // turn off the laser
+            if (usbDevice != null && usbDevice.IsOpen && hasLaser)
+                laserEnabled = false;
+
+            // ensure we're no longer acquiring
+            lock (acquisitionLock)
+            {
+                // stop all USB comms
+                shuttingDown = true;
+            }
+
+            logger.debug("Spectrometer.close: throwawaySum = {0}", throwawaySum); // just make sure it gets used
 
             if (usbDevice != null)
             {
                 if (usbDevice.IsOpen)
                 {
-                    if (hasLaser)
-                        laserEnabled = false;
-
                     IUsbDevice wholeUsbDevice = usbDevice as IUsbDevice;
                     if (!ReferenceEquals(wholeUsbDevice, null))
                         wholeUsbDevice.ReleaseInterface(0);
                     usbDevice.Close();
                 }
                 usbDevice = null;
             }
+            logger.debug($"Spectrometer.close: {id} closed");
+        }
+
+        /// <summary>
+        /// This generates a string identifier for the spectrometer which should be enough
+        /// to distinguish it from others with the same VID/PID using nothing but UsbRegistry
+        /// information (e.g. bus address).  Note that this is to be used for recognizing
+        /// previously-opened spectrometers BEFORE reading their EEPROM.
+        /// </summary>
+        /// <param name="usbRegistry"></param>
+        /// <returns></returns>
+        string generateUniqueKey(UsbRegistry usbRegistry)
+        {
+            UsbDevice device;
+            if (!usbRegistry.Open(out device))
+                return "error";
+
+            SortedDictionary<string, string> props = new SortedDictionary<string, string>();
+            foreach (KeyValuePair<string, object> pair in usbRegistry.DeviceProperties)
+            {
+                string key = pair.Key;
+                object value = pair.Value;
+
+                // we only include these properties in our key
+                if (key != "LocationInformation" &&
+                    key != "PhysicalDeviceObjectName" &&
+                    key != "Address" &&
+                    key != "Driver")
+                    continue;
+
+                if (value is string[])
+                    props[key] = string.Format("[ {0} ]", string.Join(", ", value as string[]));
+                else if (value is string)
+                    props[key] = value as string;
+            }
+            device.Close();
+
+            // flatten 
+            List<string> tok = new List<string>();
+            foreach (var pair in props)
+                tok.Add($"{pair.Key}={pair.Value}");
+            return string.Join("|", tok);

```

Comment:
Does this enumerate the spectrometer the same on a disconnect => reconnect? Pretty cool

Reply:
Basically, yeah.  That was a request from a customer who has 6+ spectrometers hooked up simultaneously, and periodically has to power-cycle one of them, and thus needs to re-call openAllSpectrometers() to re-enable the one that was bounced.  However, they don't want openAllSpectrometers() to automatically overwrite the integration time and other settings for every other spectrometer.  So the whole thing with SpectrometerUptime was basically to "guess" which spectrometer(s) may have been reset (and thus legitimately need to be initialized to defaults), versus "left alone" (all I do is log debug lines of their current settings, which has the "side-effect" of calling those settings getters, which automatically reads them from USB and caches them as the new "local value").  Not perfect, but it made things better.